### PR TITLE
Fix versionable behavior when exists 1:1 relationships

### PR DESCRIPTION
**Pull request by rafaelnery**
Fix behavior when versionable column is pk, documented methods and implementing tests. Fixing issue #1109

### CODE REVIEW

**Comment by rafaelnery** [comment]
Previously travis-ci reported fail... but re-run tests.. OK

https://travis-ci.org/propelorm/Propel2/jobs/137553549

File: src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php [hunk 1]
```diff
@@ -330,11 +365,11 @@ protected function addToVersion(&$script)
  * Sets the properties of the current object to the value they had at a specific version
  *
  * @param   integer \$versionNumber The version number to read
- * @param   ConnectionInterface \$con The connection to use
+ * @param   ConnectionInterface \$con (optional) The ConnectionInterface connection to use.
```
Comment (marcj) [inline src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php]: Can you please remove everywhere "(optional)"? We haven't used that in the code.

File: tests/Propel/Tests/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifierTest.php [hunk 2]
```diff
@@ -150,6 +150,29 @@ public static function setUpBeforeClass()
...
+        <column name="foo_id" type="integer" primaryKey="true"/>
+        <column name="bar" type="varchar" size="32"/>
+        <foreign-key foreignTable="VersionableBehaviorTestOneToOne">
+            <reference local="foo_id" foreign="id"/>
+        </foreign-key>
+        <behavior name="versionable" />
+    </table>
+</database>
+XML;
+        QuickBuilder::buildSchema($schema5);
```
Comment (marcj) [inline tests/Propel/Tests/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifierTest.php]: Only the build fails with 1:1? I don't see any real tests, only building of classes for some xml :)

File: src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php [hunk 3]
```diff
@@ -246,22 +251,49 @@ public function isVersioningNecessary(\$con = null)
...
+                $fkGetter = $this->builder->getRefFKPhpNameAffix($fk);
+                $script .= "
+    if (\$this->single{$fkGetter}) {
+
+        // to avoid infinite loops, emulate in save
+        \$this->alreadyInSave = true;
+
+        if (\$this->single{$fkGetter}->isVersioningNecessary(\$con)) {
+
+            \$this->alreadyInSave = false;
```
Comment (marcj) [inline src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php]: where exactly happens that infinite loop? You set `alreadyInSave` but never check it?

Comment (rafaelnery) [inline src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php]: Good Question :sweat_smile:

**Comment by rafaelnery** [comment]
Will correct these points tonight (here GMT -3): P

File: tests/Propel/Tests/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifierTest.php (hunk 2 above)
Comment (rafaelnery) [inline tests/Propel/Tests/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifierTest.php]: It was really where the problem occurs .. But you're right!

File: src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php (hunk 1 above)
Comment (rafaelnery) [inline src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php]: OKIDOKIE!

File: src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php (hunk 3 above)
Comment (rafaelnery) [inline src/Propel/Generator/Behavior/Versionable/VersionableBehaviorObjectBuilderModifier.php]: @marcj I searched intensively why this code ... and I could not understand ... Can you help me?

I found this Issues ... 

- #77 
- propelorm/Propel#205


In the both case(1:1 or 1:n) when remove $this->alreadyInSave...  Bang! SegFault

captura de tela de 2016-06-18 16-14-51

So the code is in loop on lines 5111 and 2006.

**Comment by ceronem** [comment]
I have the same problem in self referencing entity. It start an infinite loop in the same place of @rafaelnery.
In this case i have a self referencing entity:
$this->alredyInSave = false
$this->enforceVersion = false
ChildViaggiQuery::isVersioningEnabled() && ($this->isNew() || $this->isModified()) || $this->isDeleted() = false
and at line 15021 isVersioningNecessary($con) start the infinite loop
screen

**Comment by marcj** [comment]
Well, alright, let us merge this. Thanks!

**Pull request merged**